src/compiler/scala/tools/nsc/ast/parser/Parsers.scala

```diff
@@ -264,9 +264,13 @@ self =>
     def smartParse(): Tree = withSmartParsing {
       val firstTry = parse()
       if (syntaxErrors.isEmpty) firstTry
-      else in.healBraces() match {
-        case Nil      => showSyntaxErrors() ; firstTry
-        case patches  => withPatches(patches).parse()
+      else {
+        val patches = in.healBraces()
+        if (patches.isEmpty) {
+          showSyntaxErrors()
+          firstTry
+        }
+        else withPatches(patches).parse()
```

Comment:
still not satisfying. Maybe `in.healBraces().pipe(patches => ...)`.

Reply:
the "one-liner" is compelling? `if (patches.isEmpty) firstTry.tap(_ => showSyntaxErrors())` is expressive